snooty/test_parser.py

```diff
@@ -140,73 +141,165 @@ def test_literalinclude() -> None:
     page, diagnostics = parse_rst(
         parser,
         path,
         """
-.. literalinclude:: /driver-examples/pythonexample.py
-   :start-after: Start Example 0
-   :end-before: End Example 3
+.. literalinclude::
 """,
     )
     page.finish(diagnostics)
-    assert len(diagnostics) == 1 and isinstance(diagnostics[0], InvalidLiteralInclude)
+    assert len(diagnostics) == 1
+    assert isinstance(diagnostics[0], ExpectedPathArg)
+    check_ast_testing_string(
+        page.ast,
+        """<root>
+        <directive name="literalinclude">
+        </directive>
+        </root>""",
+    )
 
+    # Test failure to locate included code file
     page, diagnostics = parse_rst(
         parser,
         path,
         """
-.. literalinclude:: /driver-examples/pythonexample.py
-   :start-after: Start Example 3
-   :end-before: End Example 0
+.. literalinclude:: includes/does_not_exist.js
 """,
     )
     page.finish(diagnostics)
-    assert len(diagnostics) == 1 and isinstance(diagnostics[0], InvalidLiteralInclude)
+    assert len(diagnostics) == 1
+    assert isinstance(diagnostics[0], CannotOpenFile)
+    check_ast_testing_string(
+        page.ast,
+        """<root>
+        <directive name="literalinclude">
+        <text>includes/does_not_exist.js</text>
+        </directive>
+        </root>""",
+    )
 
+    # Test failure to locate start-after text
     page, diagnostics = parse_rst(
         parser,
         path,
         """
-.. literalinclude:: /driver-examples/garbagnrekvjisd.py
+.. literalinclude:: /test_parser/includes/sample_code.py
+   :start-after: does not exist
+   :end-before: end example 1
 """,
     )
     page.finish(diagnostics)
-    assert len(diagnostics) == 1 and isinstance(diagnostics[0], CannotOpenFile)
+    assert len(diagnostics) == 1
+    assert isinstance(diagnostics[0], InvalidLiteralInclude)
+    check_ast_testing_string(
+        page.ast,
+        """<root>
+        <directive name="literalinclude" start-after="does not exist" end-before="end example 1">
+        <text>/test_parser/includes/sample_code.py</text>
+        <code copyable="True">    # start example 1
+    print("test dedent")</code>
+        </directive>
+        </root>""",
+    )
+
+    # Test failure to locate end-before text
+    page, diagnostics = parse_rst(
+        parser,
+        path,
+        """
+.. literalinclude:: /test_parser/includes/sample_code.py
+   :start-after: start example 1
+   :end-before: does not exist
+""",
+    )
+    page.finish(diagnostics)
+    assert len(diagnostics) == 1
+    assert isinstance(diagnostics[0], InvalidLiteralInclude)
+    check_ast_testing_string(
+        page.ast,
+        """<root>
+        <directive name="literalinclude" start-after="start example 1" end-before="does not exist">
+        <text>/test_parser/includes/sample_code.py</text>
+        <code copyable="True">    print("test dedent")
+    # end example 1
+
+    # start example 2
+    print("hello world")</code></directive>
```

Comment:
Hmm why isn't the line `# end example 2` included here? Is it getting cut off for some reason?

Reply:
The original functionality would have cut off the last line (due to setting `end_before` to `-1`, then splicing `lines`). I've updated it such that if no `end_before` is found, it defaults to EOF (as if no `end_before` was specified).